blocksubscriptions.go

```diff
@@ -27,75 +35,138 @@ type blockMessage struct {
-			subChan = sub.onConnectBasic
-		case connectExt:
-			subChan = sub.onConnectExt
-		case disconnect:
-			subChan = sub.onDisconnect
-		default:
-			// TODO: Return a useful error when factored out into
-			// its own package.
-			panic("invalid message type")
-		}
-		if subChan != nil {
-			select {
-			case sub.notifyBlock <- bm:
-			case <-sub.quit:
-			case <-sub.intQuit:
-			}
-		}
+		sendMsgToSubscriber(sub, bm)
 	}
 	s.mtxSubscribers.RUnlock()
 }
 
+// sendMsgToSubscriber is a helper function that sends the target message to
+// the subscription client over the proper channel based on the type of the new
+// block notification.
+func sendMsgToSubscriber(sub *blockSubscription, bm *blockMessage) {
+
+	var subChan chan<- wire.BlockHeader
+
+	switch bm.msgType {
+	case connectBasic:
+		subChan = sub.onConnectBasic
+	case disconnect:
+		subChan = sub.onDisconnect
+	default:
+		// TODO: Return a useful error when factored out into its own
+		// package.
+		panic("invalid message type")
+	}
+
+	// If the subscription channel was found for this subscription based on
+	// the new update, then we'll wait to either send this notification, or
+	// quit from either signal.
+	if subChan != nil {
+		select {
+		case sub.notifyBlock <- bm:
+
+		case <-sub.quit:
+
+		case <-sub.intQuit:
+		}
+	}
+}
+
 // subscribeBlockMsg handles adding block subscriptions to the ChainService.
-// TODO(aakselrod): move this to its own package and refactor so that we're
-// not modifying an object held by the caller.
-func (s *ChainService) subscribeBlockMsg(onConnectBasic, onConnectExt,
-	onDisconnect chan<- wire.BlockHeader,
-	quit <-chan struct{}) *blockSubscription {
-	s.mtxSubscribers.Lock()
-	defer s.mtxSubscribers.Unlock()
+// The best known height to the caller should be passed in, such that we can
+// send a backlog of notifications to the caller if they're behind the current
+// best tip.
+//
+// TODO(aakselrod): move this to its own package and refactor so that we're not
+// modifying an object held by the caller.
+func (s *ChainService) subscribeBlockMsg(bestHeight uint32,
+	onConnectBasic, onDisconnect chan<- wire.BlockHeader,
+	quit <-chan struct{}) (*blockSubscription, error) {
+
 	subscription := blockSubscription{
 		onConnectBasic: onConnectBasic,
-		onConnectExt:   onConnectExt,
 		onDisconnect:   onDisconnect,
 		quit:           quit,
 		notifyBlock:    make(chan *blockMessage),
 		intQuit:        make(chan struct{}),
 	}
-	s.blockSubscribers[&subscription] = struct{}{}
-	go subscription.subscriptionHandler()
-	return &subscription
+
+	// At this point, we'll now check to see if we need to deliver any
+	// backlog notifications as its possible that while the caller is
+	// requesting right after a new set of blocks has been connected.
+	err := s.blockManager.SynchronizeFilterHeaders(func(filterHeaderTip uint32) error {
+		s.mtxSubscribers.Lock()
+		defer s.mtxSubscribers.Unlock()
+
+		s.blockSubscribers[&subscription] = struct{}{}
```

Comment:
Does this need to be done inside synchronize filter headers? Is it possible to do:
 1. create `subscription`
 2. spawn `subscriptionHandler`
3.` SychnronizeFilterHeaders`
4. If successful:
  4.1 acquire `mtxSubscribers` and add to `blockSubscribers`

Reply:
I think so, but I believe you would need to acquire `mtxSubscribers` inside `SychnronizeFilterHeaders`, since else you risk  that a new block comes in before you are able to add the new subscription.

Reply:
Yep what @halseth said.